source/lac/scalapack.cc

```diff
@@ -205,6 +205,88 @@ ScaLAPACKMatrix<NumberType>::copy_to (FullMatrix<NumberType> &matrix) const
 
 
 
+/*
+ * Copy the contents of the distributed matrix into different distributed matrix @p dest.
+ *
+ * function is supposed to also work for matrices with different process grids or block-cyclic distributions
+ */
+template <typename NumberType>
+void
+ScaLAPACKMatrix<NumberType>::copy_to (ScaLAPACKMatrix<NumberType> &dest) const
+{
+  Assert (n_rows == dest.n_rows, ExcDimensionMismatch(n_rows, dest.n_rows));
+  Assert (n_columns == dest.n_columns, ExcDimensionMismatch(n_columns, dest.n_columns));
+
+  if (this->grid->mpi_process_is_active)
+    AssertThrow (this->descriptor[0]==1,ExcMessage("Copying of ScaLAPACK matrices only implemented for dense matrices"));
+  if (dest.grid->mpi_process_is_active)
+    AssertThrow (dest.descriptor[0]==1,ExcMessage("Copying of ScaLAPACK matrices only implemented for dense matrices"));
+
+  /*
+   * just in case of different process grids or block-cyclic distributions
+   * inter-process communication is necessary
+   * if distributed matrices have the same process grid and block sizes, local copying is enough
+   */
+  if ( (this->grid != dest.grid) || (row_block_size != dest.row_block_size) || (column_block_size != dest.column_block_size) )
+    {
+      /*
+       * get the MPI communicator, which is the union of the source and destination MPI communicator
+       */
+      dealii::types::blas_int ierr = 0;
+      MPI_Group group_source, group_dest, group_union;
+      ierr = MPI_Comm_group(this->grid->mpi_communicator, &group_source);
+      AssertThrowMPI(ierr);
+      ierr = MPI_Comm_group(dest.grid->mpi_communicator, &group_dest);
+      AssertThrowMPI(ierr);
+      ierr = MPI_Group_union(group_source, group_dest, &group_union);
+      AssertThrowMPI(ierr);
+      MPI_Comm mpi_communicator_union;
+      ierr = MPI_Comm_create_group(MPI_COMM_WORLD, group_union, 5, &mpi_communicator_union);
```

Comment:
we should not hard-code `MPI_COMM_WORLD`. I think the right thing to ask is that both matrices are built upon the same mpi_communicator (`this->grid->mpi_communicator` and `dest.grid->mpi_communicator`) or not?